File: benchmarks/intensive_benchmark.js
// JavaScript Intensive Computational Benchmark
// Equivalent to Cypescript intensive benchmark

console.log("=== JavaScript Intensive Computational Benchmark ===");

// Benchmark 1: Nested Loop Computation
console.log("Benchmark 1: Nested Loop Matrix Multiplication Simulation");
const matrixSize = 200;
const matrixIterations = 100;

console.log("Matrix size: 200x200");
console.log("Iterations: 100");

let totalOperations = 0;
console.log("Starting intensive nested loop benchmark...");
const nestedStart = performance.now();

for (let iter = 0; iter < matrixIterations; iter++) {
    for (let i = 0; i < matrixSize; i++) {
        for (let j = 0; j < matrixSize; j++) {
            for (let k = 0; k < matrixSize; k++) {
                // Simulate matrix multiplication operation
                const result = (i * j + k) % 1000;
                totalOperations += result;
            }
        }
    }
}

const nestedEnd = performance.now();
const nestedTime = nestedEnd - nestedStart;
console.log("Nested loop benchmark complete");
console.log(`Total operations: ${totalOperations}`);
console.log(`Nested loop time: ${nestedTime.toFixed(2)}ms`);

// Benchmark 2: Prime Number Generation
console.log("Benchmark 2: Prime Number Generation");
const primeLimit = 10000;
let primeCount = 0;

console.log("Finding primes up to: 10,000");

console.log("Starting prime generation benchmark...");
const primeStart = performance.now();

for (let num = 2; num <= primeLimit; num++) {
    let isPrime = true;
    
    // Check if number is prime
    for (let divisor = 2; divisor * divisor <= num; divisor++) {
        if (num % divisor === 0) {
            isPrime = false;
            break;
        }
    }
    
    if (isPrime) {
        primeCount++;
    }
}

const primeEnd = performance.now();
const primeTime = primeEnd - primeStart;
console.log("Prime generation benchmark complete");
console.log(`Primes found: ${primeCount}`);
console.log(`Prime generation time: ${primeTime.toFixed(2)}ms`);

// Benchmark 3: Fibonacci Sequence
console.log("Benchmark 3: Fibonacci Sequence Generation");
const fibCount = 35;
const fibIterations = 1000;

console.log("Fibonacci numbers to calculate: 35");
console.log("Iterations: 1,000");

let fibTotal = 0;
console.log("Starting Fibonacci benchmark...");
const fibStart = performance.now();

for (let fibIter = 0; fibIter < fibIterations; fibIter++) {
    for (let n = 0; n <= fibCount; n++) {
        // Calculate nth Fibonacci number iteratively
        let a = 0;
        let b = 1;
        let fibResult = 0;
        
        if (n === 0) {
            fibResult = 0;
        } else if (n === 1) {
            fibResult = 1;
        } else {
            for (let fibI = 2; fibI <= n; fibI++) {
                fibResult = a + b;
                a = b;
                b = fibResult;
            }
        }
        
        fibTotal += fibResult;
    }
}

const fibEnd = performance.now();
const fibTime = fibEnd - fibStart;
console.log("Fibonacci benchmark complete");
console.log(`Total Fibonacci sum: ${fibTotal}`);
console.log(`Fibonacci time: ${fibTime.toFixed(2)}ms`);

// Benchmark 4: Array Sorting Simulation (Bubble Sort)
console.log("Benchmark 4: Array Sorting Simulation");
const sortSize = 1000;
const sortIterations = 50;

console.log("Array size: 1,000");
console.log("Iterations: 50");

const testArray = [64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42, 30, 18, 95, 70];
const testSize = testArray.length;
let sortOperations = 0;

console.log("Starting sorting benchmark...");
const sortStart = performance.now();

for (let sortIter = 0; sortIter < sortIterations; sortIter++) {
    // Bubble sort implementation
    for (let i = 0; i < testSize - 1; i++) {
        for (let j = 0; j < testSize - i - 1; j++) {
            if (testArray[j] > testArray[j + 1]) {
                // Swap elements
                const temp = testArray[j];
                testArray[j] = testArray[j + 1];
                testArray[j + 1] = temp;
                sortOperations++;
            }
        }
    }
    
    // Reset array for next iteration (reverse it)
    testArray.reverse();
}

const sortEnd = performance.now();
const sortTime = sortEnd - sortStart;
console.log("Sorting benchmark complete");
console.log(`Sort operations: ${sortOperations}`);
console.log(`Sorting time: ${sortTime.toFixed(2)}ms`);

// Summary
const totalTime = nestedTime + primeTime + fibTime + sortTime;
console.log("=== JavaScript Intensive Benchmark Summary ===");
console.log(`Nested loops: ${nestedTime.toFixed(2)}ms`);
console.log(`Prime generation: ${primeTime.toFixed(2)}ms`);
console.log(`Fibonacci: ${fibTime.toFixed(2)}ms`);
console.log(`Sorting: ${sortTime.toFixed(2)}ms`);
console.log(`Total time: ${totalTime.toFixed(2)}ms`);
console.log("=== Intensive Benchmark Complete ===");

File: benchmarks/intensive_benchmark.py
#!/usr/bin/env python3
# Python Intensive Computational Benchmark
# Equivalent to Cypescript intensive_benchmark.csc

import time

print("=== Python Intensive Computational Benchmark ===")

# Benchmark 1: Nested Loop Computation
print("Benchmark 1: Nested Loop Matrix Multiplication Simulation")
matrix_size = 200
matrix_iterations = 100

print(f"Matrix size: {matrix_size}x{matrix_size}")
print(f"Iterations: {matrix_iterations}")

total_operations = 0
nested_start = time.perf_counter()

for iteration in range(matrix_iterations):
    for i in range(matrix_size):
        for j in range(matrix_size):
            for k in range(matrix_size):
                # Simulate matrix multiplication operation
                result = (i * j + k) % 1000
                total_operations += result

nested_end = time.perf_counter()
nested_time = (nested_end - nested_start) * 1000

print("Nested loop benchmark complete")
print(f"Total operations: {total_operations}")
print(f"Nested loop time: {nested_time:.2f}ms")

# Benchmark 2: Prime Number Generation
print("Benchmark 2: Prime Number Generation")
prime_limit = 10000
prime_count = 0

print(f"Finding primes up to: {prime_limit:,}")

prime_start = time.perf_counter()

for num in range(2, prime_limit + 1):
    is_prime = True
    
    # Check if number is prime
    for divisor in range(2, int(num**0.5) + 1):
        if num % divisor == 0:
            is_prime = False
            break
    
    if is_prime:
        prime_count += 1

prime_end = time.perf_counter()
prime_time = (prime_end - prime_start) * 1000

print("Prime generation benchmark complete")
print(f"Primes found: {prime_count}")
print(f"Prime generation time: {prime_time:.2f}ms")

# Benchmark 3: Fibonacci Sequence Generation
print("Benchmark 3: Fibonacci Sequence Generation")
fib_count = 35
fib_iterations = 1000

print(f"Fibonacci numbers to calculate: {fib_count}")
print(f"Iterations: {fib_iterations:,}")

fib_total = 0
fib_start = time.perf_counter()

for iteration in range(fib_iterations):
    for n in range(fib_count + 1):
        # Calculate nth Fibonacci number iteratively
        if n == 0:
            fib_result = 0
        elif n == 1:
            fib_result = 1
        else:
            a, b = 0, 1
            for i in range(2, n + 1):
                fib_result = a + b
                a, b = b, fib_result
        
        fib_total += fib_result

fib_end = time.perf_counter()
fib_time = (fib_end - fib_start) * 1000

print("Fibonacci benchmark complete")
print(f"Total Fibonacci sum: {fib_total}")
print(f"Fibonacci time: {fib_time:.2f}ms")

# Benchmark 4: Array Sorting Simulation (Bubble Sort)
print("Benchmark 4: Array Sorting Simulation")
sort_size = 1000
sort_iterations = 50

print(f"Array size: {sort_size}")
print(f"Iterations: {sort_iterations}")

test_array = [64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42, 30, 18, 95, 70]
test_size = len(test_array)
sort_operations = 0

sort_start = time.perf_counter()

for iteration in range(sort_iterations):
    # Bubble sort implementation
    arr = test_array.copy()
    for i in range(test_size - 1):
        for j in range(test_size - i - 1):
            if arr[j] > arr[j + 1]:
                # Swap elements
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                sort_operations += 1

sort_end = time.perf_counter()
sort_time = (sort_end - sort_start) * 1000

print("Sorting benchmark complete")
print(f"Sort operations: {sort_operations}")
print(f"Sorting time: {sort_time:.2f}ms")

# Summary
total_time = nested_time + prime_time + fib_time + sort_time
print("=== Python Intensive Benchmark Summary ===")
print(f"Nested loops: {nested_time:.2f}ms")
print(f"Prime generation: {prime_time:.2f}ms")
print(f"Fibonacci: {fib_time:.2f}ms")
print(f"Sorting: {sort_time:.2f}ms")
print(f"Total time: {total_time:.2f}ms")
print("=== Intensive Benchmark Complete ===")
